# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: media/mtransport/ipc/WebrtcTCPSocket.cpp
# Commit: b879a0044cf3
# Full Hash: b879a0044cf39f0230879d075c915485affb7707
# Author: Byron Campen [:bwc] <docfaraday@gmail.com>
# Date: 2019-09-18 21:49:27
# Regressor Bug: 1569183
# File Overlap Count: 2
# Description:
#   Bug 1569183: Teach WebrtcTCPSocket to bind to a specific host/port. r=mjf
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D45102
# ==============================================================================

diff -r f5353ea9f1df -r b879a0044cf3 media/mtransport/ipc/WebrtcTCPSocket.cpp
--- a/media/mtransport/ipc/WebrtcTCPSocket.cpp	Tue Sep 17 18:17:14 2019 +0000
+++ b/media/mtransport/ipc/WebrtcTCPSocket.cpp	Tue Sep 17 18:17:18 2019 +0000
@@ -137,6 +137,8 @@
 }
 
 nsresult WebrtcTCPSocket::Open(const nsCString& aHost, const int& aPort,
+                               const nsCString& aLocalAddress,
+                               const int& aLocalPort,
                                const net::LoadInfoArgs& aArgs,
                                const nsCString& aAlpn,
                                NrSocketProxyConfig::ProxyPolicy aProxyPolicy) {
@@ -166,6 +168,8 @@
   if (aProxyPolicy == NrSocketProxyConfig::kForceProxy) {
     mForceProxy = true;
   }
+  mLocalAddress = aLocalAddress;
+  mLocalPort = aLocalPort;
 
   if (aProxyPolicy == NrSocketProxyConfig::kDisableProxy) {
     rv = OpenWithoutHttpProxy(nullptr);
@@ -278,6 +282,26 @@
     return rv;
   }
 
+  mTransport->SetReuseAddrPort(true);
+
+  PRNetAddr prAddr;
+  if (NS_WARN_IF(PR_SUCCESS !=
+                 PR_InitializeNetAddr(PR_IpAddrAny, mLocalPort, &prAddr))) {
+    return NS_ERROR_FAILURE;
+  }
+
+  if (NS_WARN_IF(PR_SUCCESS !=
+                 PR_StringToNetAddr(mLocalAddress.BeginReading(), &prAddr))) {
+    return NS_ERROR_FAILURE;
+  }
+
+  mozilla::net::NetAddr addr;
+  PRNetAddrToNetAddr(&prAddr, &addr);
+  rv = mTransport->Bind(&addr);
+  if (NS_WARN_IF(NS_FAILED(rv))) {
+    return rv;
+  }
+
   nsCOMPtr<nsIInputStream> socketIn;
   rv = mTransport->OpenInputStream(0, 0, 0, getter_AddRefs(socketIn));
   if (NS_WARN_IF(NS_FAILED(rv))) {